### DISCUSSION TITLE
Proposal of new TApp vs host business model in epoch4

### DISCUSSION THREAD

**Discussion by kevingzhang** [tearust/teaproject#69] [Ideas]
# Problems of TApp vs Host business model in epoch3
In epoch3, sudo is the single role to create TApps. Sudo needs to pay a huge initial fee and daily tipping to keep the TApps from bankruptcy. This is not a sustainable model. In Epoch4, the TApps are created by any contestants, I doubt this model will make any tApps profitable.

# Change proposals in epoch4

## Anyone can create TApps.

Sudo will not create tapps. Anyone else can do. 

## A new Twitter TApp added, alone with existing (improved) Youtuber tapp

Besides Youtuber TApp, we will add a new twitter TApp. Anyone can create a tweet link as the content of this tapp. The retweet count will be used to evaluate this TApp to receive tipping. The retweet is the views in the Youtuber tapp.

Similar Reddit TApp may be added too

## Creator is the owner of the TApp
Whoever creates a TApp is the owner. The owner can decide the host billing model and pricing. Of course, a initial funding is required to start a TApp. The owner also take  30% bonding curve incentive. See more on bonding curve concept.

# A new host billing model added

Besides the existing fixed performance -> host fee model, there will be a new more competitive model to be added.

## Staking tokens

In this new model, the hosting miners share the profit and risk with the TApp owners. There is no fixed income from hosting service. The host need to use the CML's performance to host an TAp, in return, he gets staking TApp tokens. These tokens are staked. Host miners can receive profit sharing (consuming profit). But as long as the CML is still hosting, the staking 
the token cannot be sold. 

## Min and max hosting CMLs
There are min and max hosts. If the hosting CML number is lower than the min hosts, the TApp won't start. If the hosting CML number reaches the max, no more CML will be accepted. The min number is fixed to 3. This is the minimal requirement for RAFT consensus.

## No expense, no token amount shrinking
In the old business model, the TApp automatically sell the TApp token for T to pay the fixed hosting fee. This causes the holding token to shrink over time. If the TApp did not win the tipping competition, the investor's TApp token will continue shrinking. Eventually bankruptcy. The investor lost everything.

In this new model, the hosting CML shares risk with the TApp owners(investors), there will not be any hosting fee to be paid to the hosting CMLs. So no expense, no reduction in the TApp token number for regular investors.

## TApp owners can still choose the old billing model, as long as he can afford the cost of hosting

The new and old billing models can coexist until no one actually chooses the old model. 

## Hosting miners income
Instead of the fixed mining fee. the hosting miners receive the dividend when tipping as epoch 3 does. The sudo NPC still pay tipping based on the rewarding rules. The tips will increase the total supply of token. The new increased token distributed to all token holders, including the miners. because the miners owns the staked token.

## Unhost
If hosting miners decide stop hosting this TApp. He can unhost. The performance number will be returned to this CML's remaining performance, the staking token will be burned. This may leave room from the max host number so that new CML to start hosting.

## Bankruptcy
As long as the hosting miners number above the min requirement (we hard coded it to 3), the TApp will not die. 
but if hosting miners think there is no benefit by hosting this TApp, they decided to unhost. as long as the remaining hosting CML number less than the min requirement. the TApp will no longer be available. It is considered "dead".

# Creating new TApp
The owner will choose the following parameters when creating new tapps
- Max CML hosts.  This number has to be greater than the min requirement (3). But if it is too big, less dividend is shared to all investors, including the owner himself. I would suggest 5 or 6.
- Min CML hosts. this is fixed number  3. When there is only less than 3 hosts, the TApp cannot start.
- Host performance requirement. This is how much the performance reduced from the CML's remaining performance. We fix this number to 3000.
- Every hosting CML received token number.  This number is the TApp tokens issued to the hosting CML. The higher this number, the more attractive to the miners. If this number is too low, miners would like to host other TApps else.

# Topic list
Everyone can create TApps, but the topic (the youtube links or tweet links) has to be in a sudo list.
Every link can be used by only one TApp. If all links are used to other TApps, you have to wait till any of them die or new link added to the list. 

All topic in the list has to be approved by sudo, that related and promoting TEA project in social media.

Anyone can propose new links in Telegram channel. If it gets proof, the proposer has the priority to create this TApp.

# Other changes related to the CML

The entry level will be increase to own and plant a CML. This will reduce the number of mining CML. We encourage those who cannot afford a CMLto invest or staking instead. In the real world, only professional miners can afford mining.

**Comment by geopor** [comment]
> **Anyone can create TApps.**

Would suggest the number for every person be limited, else risk network-clogging spam with mass-created TApps.

> **Twitter TApp added, alone with existing (improved) Youtuber tapp**
- Twitter TApps compete against Twitter TApps, winners for the big **consume** action for the day I would suggest using the number of retweets because they'll be a relatively small number instead of % increase.
- YouTube TApps compete against YouTube TApps, daily **consume** winners still based on % increase in views.
- Reddit TApps compete against Reddit TApps based on number of upvotes. Would also suggest using number instead of % increase as they'll be smaller in size.

The consume action (tip) is proposed to be spread out over a greater number of TApps -> https://github.com/tearust/teaproject/discussions/68#discussioncomment-1329423

> **Creator is the owner of the TApp**
> The owner can decide the host billing model and pricing. Of course, a initial funding is required to start a TApp. The owner also take 30% bonding curve incentive.

Since this TEA outlay could be financially burdensome and prevent them from taking part in other aspects of the contest, I would suggest a special TApp creator's loan be available. It can be an option when someone creates a new TApp - **Do you want to take out a loan to help fund this TApp's bonding curve?** It should be minimal interest similar to Epoch 2.1 TEA Genesis loan interest rate (~ .1% every 3 days). And it wouldn't be an arbitrary amount, it would be solely to fund the TApp's bonding curve.

> **The entry level will be increase to own and plant a CML.**

I already thought it was expensive to mine but if the goal is to reduce the number of miners that will have the intended effect. I still hope everyone that joins our contest can mine if they are technically up to it, I don't think cost should be a deterrent since in real life, RPi or DigitalOcean is fairly cheap to run.

**Reply by kevingzhang** [reply]
> Would suggest the number for every person be limited, else risk network-clogging spam with mass-created TApps.

The number of TApps is limited by
- the links in the approval links list. Only links in this list can be used to create a new TApp. 
- creating TApp need to pay initial fund. There is only one chance to input the initial funding. If the creator did not put the max number he has, he misses the single chance. Because the price is lowest at that point.

**Comment by kevingzhang** [comment]
> Twitter TApps compete against Twitter TApps, winners for the big consume action for the day I would suggest using the number of retweets because they'll be a relatively small number instead of % increase.
If some tweets have a large retweet already, if we do not calculate the increase rates, the other tweets will lose the chance to win forever. 
The same applies to upvotes in Reddit. If everyone is clear on who is the winner, no one else would like to join the game.

This is the reason why I suggested to have two tips. One for increase rate, another for absolute number.

**Comment by kevingzhang** [comment]
> Since this TEA outlay could be financially burdensome and prevent them from taking part in other aspects of the contest, I would suggest a special TApp creator's loan be available. It can be an option when someone creates a new TApp - Do you want to take out a loan to help fund this TApp's bonding curve? It should be minimal interest similar to Epoch 2.1 TEA Genesis loan interest rate (~ .1% every 3 days). And it wouldn't be an arbitrary amount, it would be solely to fund the TApp's bonding curve.

This is a good idea. We can do this to help the create to start the TApp.

**Reply by kevingzhang** [reply]
> > Since this TEA outlay could be financially burdensome and prevent them from taking part in other aspects of the contest, I would suggest a special TApp creator's loan be available. It can be an option when someone creates a new TApp - Do you want to take out a loan to help fund this TApp's bonding curve? It should be minimal interest similar to Epoch 2.1 TEA Genesis loan interest rate (~ .1% every 3 days). And it wouldn't be an arbitrary amount, it would be solely to fund the TApp's bonding curve.
> 
> This is a good idea. We can do this to help creators to start the TApp.

There is one technical difficulty. The current genesis loan can only apply to CML seeds. I do not know what the TApp creator can use as collateral when borrowing Genesis loan. What is on your mind?

**Reply by geopor** [reply]
Just like miners get stake tokens they cannot touch, TApp creators could put up the TApp's tokens as collateral for the stake that they can't sell until they pay off the loan. Paying off the loan releases the "staked" tokens back into the normal section of the wallet.

**Comment by kevingzhang** [comment]
> I already thought it was expensive to mine but if the goal is to reduce the number of miners that will have the intended effect. I still hope everyone that joins our contest can mine if they are technically up to it, I don't think cost should be a deterrent since in real life, RPi or DigitalOcean is fairly cheap to run.

I am thinking epoch4 is a ramp up to epoch5 or 6 which setup own virtual machine to plant A CML is required. Since there is actually a cost and technical barrier, no doubt it is harder than today. Since we cannot have such a barrier simulation in epoch4, the only thing in my mind is to increase the cost as barrier. If you have any better idea to mimic the technical barrier, we can consider it too.

**Reply by geopor** [reply]
One idea is to have a hosted shell command line which would require some interaction and require at least a base line of unix command line skills. After running the user runs the (simulated) shell commands, if they did it correctly the shell would output a number which they would have to input into the mining modal to "unlock" the mining machine. 

We wouldn't need to check the exact number, it would just need to follow a format to be accepted. Would be similar to the sign up page rejecting a sign up if they don't use proper polkadot address format.

**Reply by kevingzhang** [reply]
It is far more complicated than just typing a few linux shell command.

**Reply by geopor** [reply]
The idea is not to make it as complicated, but just enough of a technical hurdle to get people used to using the shell. It would also give linux rookies a chance to learn some beginning steps. 

Another idea is that instead of making it more expensive, we have shell commands that they can execute to get a discount on mining. If they don't know how to run the commands, then they pay full price. In this scenario, mining would only be more expensive for those who will likely have difficulty with it in epoch 5 or 6.

**Comment by geopor** [comment]
> **(miners unhost)**: The performance number will be returned to this CML's remaining performance, the staking token will be burned.

If the token is burned, 100% of the resulting TEA goes to the creator?

There is also the case of excessive unhosts which would burn too much of the staked token supply leaving too few for miners.

**Reply by kevingzhang** [reply]
Burn is simply burn, the result is ash and CO2, there will not be anything return to the creator. I am not sure where the TEA you mentioned coming from.

**Reply by kevingzhang** [reply]
I do not understand the second sentence either. Please explain.

**Reply by geopor** [reply]
When the owner creates the TApp, they pay for the initial tokens as well as the staked tokens to be distributed to miners. My question is let's say there are 100 stake tokens and a maximum of 10 hosts.

- Does every miner get 10 tokens set aside for them when they host i.e. if there are only 7 miners, then only 70 tokens are assigned to them and 30 remain in the stake pool?
- Let's imagine all 10 mining slots are taken, each miner gets 10 stake tokens. If a miner unstakes, then do those 10 stake tokens return to the stake pool? There is now 1 open mining slot and that new miner will need to receive stake tokens.

**Reply by kevingzhang** [reply]
> When the owner creates the TApp, they pay for the initial tokens as well as the staked tokens to be distributed to miners. My question is let's say there are 100 stake tokens and a maximum of 10 hosts.

Why did the creator pay for the staked tokens to be distributed to miners? I did not know this. does it happen? The distributed token to miners are "minted" not bought by the creator.

> 
> * Does every miner get 10 tokens set aside for them when they host i.e. if there are only 7 miners, then only 70 tokens are assigned to them and 30 remain in the stake pool?

30 tokens are not minted yet.
Make sure you know what mint means.


> * Let's imagine all 10 mining slots are taken, each miner gets 10 stake tokens. If a miner unstakes, then do those 10 stake tokens return to the stake pool? There is now 1 open mining slot and that new miner will need to receive stake tokens.

not return to the stake pool, they are burnt. There is no staking pool. there is NO staking pool

**Comment by kevingzhang** [comment]
![image](https://user-images.githubusercontent.com/1761809/135147580-3034323a-ce8c-41ff-a029-c126ee5f7da3.png)
The owner do not pay for the staked tokens to be distributed to miners. The
owners pay for the initial tokens to themselves. Miners gain tokens by
providing hosting services.

In the case you mentioned, the owner receives whatever he set in the
"initial token" input box, each miner receive 100 token as shown in the
screenshot above, max 10 miners allowed.
If there are 10 miners host this tapp, total 1000 (assume this the owner
init token) + 100*10 = 2000 tokens are issued
…

**Reply by geopor** [reply]
So these extra 1000 staked tokens are like "ghost" tokens? Meaning they can never be sold (because nobody paid for them and the bonding curve can't pay out what was never funded), they only act as claims for a portion of the consume action dividends.

And the TApp creator will always be able to give out these 100 stake tokens in the example above - if the 10th miner unhosts, their staked tokens are burnt but at the same time the TApp has 100 of these stake tokens to allocate to the next miner?

**Reply by kevingzhang** [reply]
Yes, those token cannot be sold by anyone. It can only be burnt when unhost. So those 1000 token is totally out of the scope of the reservation pool. it is just some kind of collateral but still gain dividends.
Yes and no. the 100 token is always ready to allocate to the next miners, but since it is burnt, the total supply reduced 100, and no dividend will be given. 

Totalsupply = hosting token + purchase token.
Purchase token can be sold and anyone
hosting token can be burned or mint for miners.
both of them earn dividends.
The totalsupply will affect the token price, so the hosting token too.

**Reply by geopor** [reply]
That token design pushes the entire bonding curve over to the right absent any economic activity. In the test wallet I can see TApp tokens with less than 50 outstanding that have prices above what we saw in epoch 3 for TApp tokens with over 1000 outstanding tokens.

The expense isn't clear to TApp creators - here is someone thinking they'll pay a modest amount of TEA when they select 100 initial tokens:

![Screen Shot 2021-09-28 at 12 13 59 PM](https://user-images.githubusercontent.com/86096370/135159859-e0721b80-a1cc-4da3-a7b1-a7c5bb2f4b2c.png)

Just as a test, I selected 1000 token share, and that ended up taking over 6000T out of the TApp creators wallet.

My suggestion would be either to make the cost transparent to the TApp creator as the **Calculate cost** button is misleading since it's counting from zero and not the bonding curve puffed up with the staked tokens. Another possibility is to just make it non-selectable and done through a formula, something like **staked tokens per miner** = **initial tokens** / **max number of miners**.

This also takes care of the problem of having to list **Staked tokens / miner** in the TApps list as miners would otherwise want to shop around for most generous TApps.

**Reply by geopor** [reply]
Also I tried different amounts when creating the quantity for each staking slot - 10, 1000, 5000 - and they all add the same amount to the cost of creating the TApp in addition to the initial token cost - 6000T.

**Reply by kevingzhang** [reply]
> That token design pushes the entire bonding curve over to the right absent any economic activity. In the test wallet I can see TApp tokens with less than 50 outstanding that have prices above what we saw in epoch 3 for TApp tokens with over 1000 outstanding tokens.
> 
> The expense isn't clear to TApp creators - here is someone thinking they'll pay a modest amount of TEA when they select 100 initial tokens:
> 
> ![Screen Shot 2021-09-28 at 12 13 59 PM](https://user-images.githubusercontent.com/86096370/135159859-e0721b80-a1cc-4da3-a7b1-a7c5bb2f4b2c.png)
> 
> Just as a test, I selected 1000 token share, and that ended up taking over 6000T out of the TApp creators wallet.

I see it, and what is the problem? If creator want to own 1000 initial token, he need to pay that much of TEA. If he doesn't have so much, he can choose a much smaller initial token, such as  10 or  100. He choose  1000 may cause he has a big owner share but discourage other investors to invest in. He should know this consequence.


> 
> My suggestion would be either to make the cost transparent to the TApp creator as the **Calculate cost** button is misleading since it's counting from zero and not the bonding curve puffed up with the staked tokens. Another possibility is to just make it non-selectable and done through a formula, something like **staked tokens per miner** = **initial tokens** / **max number of miners**.

Please read my previous answer, you seem not clear about mint and burn. there is no such miner's token pool, there is no cost to the creator, the creator pay nothing to miners. Creators and miners are partner, they are not making a buy/sell deal
> 
> This also takes care of the problem of having to list **Staked tokens / miner** in the TApps list as miners would otherwise want to shop around for most generous TApps.

**Reply by kevingzhang** [reply]
> Also I tried different amounts when creating the quantity for each staking slot - 10, 1000, 5000 - and they all add the same amount to the cost of creating the TApp in addition to the initial token cost - 6000T.

Are you saying when you put the initial token as 10,  1000,  5000, they all require 6000T? if so, create an issue, this must a bug

**Reply by kevingzhang** [reply]
> > Also I tried different amounts when creating the quantity for each staking slot - 10, 1000, 5000 - and they all add the same amount to the cost of creating the TApp in addition to the initial token cost - 6000T.
> 
> Are you saying when you put the initial token as 10, 1000, 5000, they all require 6000T? if so, create an issue, this must a bug

I tried in the test wallet, I input  10,100,1000, but the number in TEA I got was not the same. they are not 6000T. they are different
do you need a screenshot?

**Comment by geopor** [comment]
After hammering out some finer details about the exact nature of these staked miner tokens, here's a TLDR summary:

 **Staked tokens per miner** is an amount allocated to every miner who hosts the TApp. These tokens aren't like normal tokens: they can't be sold and they don't augment the supply curve (and thus don't affect the TApp token price). Additionally, the TApp creator doesn't have to pay for them, they only have to select how many of these staked tokens they want provisioned for each hosting miner. If **100** staked tokens are selected per miner, then miners will receive consume action dividends _as if they owned 100 tokens_ every 1000 blocks.

These staked tokens will show up in the miners wallet under **My hosting TApps**:

![Screen Shot 2021-09-30 at 4 21 12 PM](https://user-images.githubusercontent.com/86096370/135542266-16c3ab72-0de6-4d22-b71e-989297a023b8.png)

Note that the number of staked tokens will be listed under **My hosting token**, and the total hosting income they've generated to date will be under **Total hosting token income**. Again they can't be sold and don't affect the total token supply; they are designed to give miners a place at the table anytime consume action rewards are served to token holders.
